pkg/cmd/roachtest/roachtestutil/mixedversion/planner_test.go

```diff
@@ -959,97 +970,112 @@ func Test_ChooseUpgradePathInvariants(t *testing.T) {
-
-		// The minBootstrapVersion can be any version less than or equal to minSupportedVersion.
-		minBootstrapVersionOffset := 0
-		if upgradesRemaining > 0 {
-			minBootstrapVersionOffset = rng.Intn(upgradesRemaining)
+		if msv.Version.Equals(finalVersion.Version) {
+			// Chosen minimumSupportedVersion equals finalVersion, assume it's unspecified.
+			values[2] = reflect.ValueOf("")
+		} else {
+			values[2] = reflect.ValueOf(msv.Version.String())
 		}
-		upgradesRemaining -= minBootstrapVersionOffset
-		minBootstrapVersion := randomNthPredecessor(rng, minSupportedVersion, minBootstrapVersionOffset)
-		values[2] = reflect.ValueOf(minBootstrapVersion.Version.String())
 
 		// Find the maximum amount of possible upgrades from the minBootstrapVersion
 		// and pick a random value [1, maxUpgradesFromMBV] to set numUpgrades to.
-		maxUpgradesFromMBV, err := release.MajorReleasesBetween(&finalVersion.Version, &minBootstrapVersion.Version)
+		maxUpgradesFromMBV, err := release.MajorReleasesBetween(&finalVersion.Version, &initialVersion.Version)
 		require.NoError(t, err)
 
 		numUpgrades := rng.Intn(maxUpgradesFromMBV) + 1
 		values[3] = reflect.ValueOf(numUpgrades)
-
 		values[4] = reflect.ValueOf(rng.Float64() > 0.5)
 	}
 
-	cfg := quick.Config{
-		MaxCount: 100,
-		Rand:     rng,
-		Values:   generator,
-	}
-
-	var fatalErr error
-	fatalFunc := func() func(...interface{}) {
-		fatalErr = nil
-		return func(args ...interface{}) {
-			require.Len(t, args, 1)
-			err, isErr := args[0].(error)
-			require.True(t, isErr)
-
-			fatalErr = err
-		}
-	}
+	// Counts the total number of skip upgrades across all generated test plans.
+	totalSkips := 0
 
 	verifyPlan := func(
-		finalVersion string, minSupportedVersion string, minBootstrapVersion string, numUpgrades int, skipVersions bool,
+		finalVersion string, minBootstrapVersion string, minSupportedVersion string, numUpgrades int, skipVersions bool,
 	) bool {
 		// The top level withTestBuildVersion will take care of resetting this for us.
 		_ = withTestBuildVersion(finalVersion)
 
-		// Set up our test plan using the generated values.
-		mvt := newTest(
-			NumUpgrades(numUpgrades),
-			MinimumSupportedVersion(minSupportedVersion),
-			MinimumBootstrapVersion(minBootstrapVersion),
-		)
-		if skipVersions {
-			mvt.options.skipVersionProbability = 1
+		genTest := func() *Test {
+			// Set up our test plan using the generated values.
+			opts := []CustomOption{NumUpgrades(numUpgrades)}
+			if minBootstrapVersion != "" {
+				opts = append(opts, MinimumBootstrapVersion(minBootstrapVersion))
+			}
+			if minSupportedVersion != "" {
+				opts = append(opts, MinimumSupportedVersion(minSupportedVersion))
+			}
+			mvt := newTest(
+				opts...,
+			)
+			if skipVersions {
+				mvt.options.skipVersionProbability = 1
+			} else {
+				mvt.options.skipVersionProbability = 0
+			}
+			// Setup user hooks so that we can exercise `mvt.assertExpectedUserHooks`.
```

Comment:
Added user hooks so that we can also exercise a new assertion in the planner, i.e., `assertExpectedUserHooks`.